### ISSUE TITLE
Ignoring ZSTs in `check_distinct` is unsound

### ISSUE THREAD

**Issue by wackbyte** [matthieu-m/ghost-cell#32]
Since certain ZSTs in libraries have special capabilities (such as `GhostToken`), allowing `borrow_mut` to provide mutable references to them without limit is unsound.

Example:
```rs
GhostToken::new(|mut outer| {
    GhostToken::new(|inner| {
        let inner = GhostCell::new(inner);
        let (inner1, inner2) = (&inner, &inner).borrow_mut(&mut outer).unwrap();

        let cell = GhostCell::new(false);
        let cell1 = cell.borrow_mut(inner1);
        let cell2 = cell.borrow_mut(inner2);
        *cell1 = true;
    })
})
```

Miri:
```rust
error: Undefined Behavior: attempting a write access using <241433> at alloc62629[0x0], but that tag does not exist in the borrow stack for this location
 --> <source>:9:9
  |
9 |         *cell1 = true;
  |         ^^^^^^^^^^^^^
  |         |
  |         attempting a write access using <241433> at alloc62629[0x0], but that tag does not exist in the borrow stack for this location
  |         this error occurs as part of an access at alloc62629[0x0..0x1]
  |
  = help: this indicates a potential bug in the program: it performed an invalid operation, but the Stacked Borrows rules it violated are still experimental
  = help: see https://github.com/rust-lang/unsafe-code-guidelines/blob/master/wip/stacked-borrows.md for further information
help: <241433> was created by a Unique retag at offsets [0x0..0x1]
 --> <source>:7:21
  |
7 |         let cell1 = cell.borrow_mut(inner1);
  |                     ^^^^^^^^^^^^^^^^^^^^^^^
help: <241433> was later invalidated at offsets [0x0..0x1] by a Unique retag
   --> src/ghost_cell.rs:177:18
    |
177 |         unsafe { &mut *self.value.get() }
    |                  ^^^^^^^^^^^^^^^^^^^^^^
    = note: BACKTRACE (of the first span):
    = note: inside closure at <source>:9:9: 9:22
```

(I considered this issue when thinking about possible fixes for #31 but never mentioned it, so sorry 😅)

**Comment by matthieu-m** [comment]
That's... cunning.

I focused on the memory implications, and was happy that Miri agreed with me that an overlapping ZST could be borrowed mutably multiple times concurrently, but never considered special capabilities.

I suppose elaborate work arounds could be pulled off based on whether the ZST is `Copy` or not -- as a `Copy` one cannot possibly encode a capability -- but at this point it doesn't feel worth it.

I'm thinking of reverting to the previous behavior (pre #31 ) and adding documentation that even though ZSTs are technically never overlapping they will be considered overlapping for the purpose of `GhostBorrowMut` due the aforementioned "special" capabilities they may be encoding.

Vexing, but soundness trumps all.

**Comment by matthieu-m** [comment]
Fixed in v0.2.6 which reverts the patch introduced in v0.2.5, swaps the tests introduced in v0.2.5 to guarantee detection of the cases, and adds documentation in the module explaining why it's a bad idea.

I also yanked v0.2.5 for good measure. Automated updates can pick up v0.2.6 so this shouldn't be too disruptive.
